kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,591 @@
+            -- remove from answers
+            table_remove(answers, i)
+        end
+    end
+
+    return unmatched
+end
+
+
+local function process_answers_fields(self, answers)
+    local errcode = answers.errcode
+    if not errcode then
+        local ttl = answers[1].ttl
+
+        for _, answer in ipairs(answers) do
+            -- A compromise regarding https://github.com/Kong/kong/pull/3088
+            if answer.type == TYPE_AAAA then
+                answer.address = utils.ipv6_bracket(answer.address)
+            elseif answer.type == TYPE_SRV then
+                answer.target = utils.ipv6_bracket(answer.target)
+            end
+
+            ttl = math_min(ttl, answer.ttl)
+        end
+
+        answers.ttl = self.valid_ttl or ttl
+
+    elseif errcode == 3 or errcode == 101 then
+        answers.ttl = self.empty_ttl
+
+    else
+        answers.ttl = self.error_ttl
+    end
+end
+
+
+-- NOTE: it might insert unmatched answers into cache
+local function process_answers(self, qname, qtype, answers)
+    if not answers.errcode then
+        local unmatched = filter_unmatched_answers(qname, qtype, answers)
+        for k, a in pairs(unmatched) do
+            process_answers_fields(self, a)
+            self.cache:set(k, { ttl = a.ttl }, a)
+            if not get_last_type(self.cache, a[1].name) then
+                insert_last_type(self.cache, a[1].name, a[1].type)
+            end
+        end
+
+        if #answers == 0 then
+            answers.errcode = 101
+            answers.errstr = client_errors[101]
+        end
+    end
+
+    process_answers_fields(self, answers)
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+    --log(tries, "query")
+
+    local key = name .. ":" .. qtype
+    stats_count(self.stats, key, "query")
+
+    local r, err = resolver:new(self.r_opts)
+    if not r then
+        return nil, "failed to instantiate the resolver: " .. err
+    end
+
+    local options = { additional_section = true, qtype = qtype }
+    local answers, err, q_tries = r:query(name, options, {})
+    if r.destroy then
+        r:destroy()
+    end
+
+    if not answers then
+        stats_count(self.stats, key, "query_fail")
+        return nil, "Query to DNS server failed: " .. (err or "unknown")
+    end
+
+    process_answers(self, name, qtype, answers)
+
+    stats_count(self.stats, key, answers.errstr and
+                                 "query_err:" .. answers.errstr or "query_succ")
+
+    --log(tries, answers.errstr or #answers)
+
+    return answers, nil, answers.ttl
+end
+
+
+local function start_stale_update_task(self, key, name, qtype)
+    stats_count(self.stats, key, "stale")
+
+    timer_at(0, function (premature)
+        if premature then return end
+
+        local answers = resolve_query(self, name, qtype, {})
+        if answers and (not answers.errcode or answers.errcode == 3) then
+            self.cache:set(key, { ttl = answers.ttl }, answers)
```

Comment:
Ditto, using `mlcache:set()` should also be considered a strong anti-pattern. The proper usage of this library is to only rely on `mlcache:get()`, as all other write operations require heavy lifting with `:update()` (which requires IPC to take place). This is all documented in the mlcache docs. These functions are for convenience for development purposes *or* for users who are extremely confident about what they are doing for specific use-cases. In the case of a DNS client, I do not think any other function but `:get()` should be necessary. `:get()` the most pure, clean, and powerful usage of mlcache (along with `:get_bulk()`).

Reply:
Yes, I also don't like to update some key-value via `mlcache:set()` in the callback of `mlcache:get()`. But there are some scenarios when when we use `mlcache:get()` to update the cache, we also want to update some associated item/status in the cache, I dont know how to update them without `mlcache:set()`. There two scenarios here:
1. when we update a new dns record via network query in the callback of `mlcache:get()`, we also want to update the last successful query type for this key(domain name)
2. currently we will seperate the received dns record according its domain name, for example, if we resolve x.com and get result `x.com -> cname: y.com` and `y.com -> 1.1.1.1`, the kv <`x.com`:`y.com`> will be auto insert into the mlcache by returning it from the `mlcache:set()` callback, while the <`y.com`:`1.1.1.1`> will be mannually set via `mlcache:set()`

However, you also reminded me that it can be avoided from a business perspective. For example, in the second scenario, I can refrain from splitting the intermediate results of caching DNS CNAME and directly store the final result after the CNAME reference. This way, there will only be one key-value pair to store, and there is no need for a `mlcache:set()` operation.

Reply:
It seems to me in this use case we could maybe use `:set()` without `:update()`? That would be ideal, because IPC is only necessary for L1 cache invalidation but is also hugely detrimental to performance and stability (unless the IPC is rock-solid). mlcache was built for the caching layering, and leaves the IPC problem to be solved by another library, hence why the off-the-shelf solution: it allows for testing of mlcache, but users should use their own IPC library.

Anyway, in this case, is there a situation where we need to invalidate domain names from the L1 cache, or can we simply rely on their `ttl` to be invalidated instead? Because if the former, then there is no need for `:update()` at all.

1. Query for `x.com` in worker 1 -> not in L1, not in L2
2. L3 callback for `x.com` -> `cname: y.com`
3. L3 callback for `y.com` -> `A: 1.1.1.1`
  a. In the same L3 callback, `:set("x.com", "1.1.1.1")` -> now `x.com = 1.1.1.1` is in L2.
4. Query for `x.com` in worker 2 -> L2 hit with `x.com = 1.1.1.1`
5. Eventually, `ttl` for `x.com` is reached, and `x.com = 1.1.1.1` expires in L2 and expires in all workers' L1 cache as well.

In the above scenario, there is no need for L1 cache invalidation because DNS records always expire via TTL. Is this acceptable? If so, that is how proper mlcache usage should be designed: avoiding the need for `:update()` calls at all cost, unless absolutely necessary *and* supported by a strong IPC library.

Reply:
Yes, I also want to avoid `:update()`. 

> In the above scenario, there is no need for L1 cache invalidation because DNS records always expire via TTL. Is this acceptable?

Yes, I think so, and I'll also review whether we dont need to call `:update()` for the other situation when we call `:set()` mannually . 

And for this code line, I think I have found another good way to avoid calling `:set()`, I just call `:get()` with a new simple callback to query and update the stale/expired record.  See https://github.com/Kong/kong/pull/12305/commits/ff8329f4095b071b7b1d2aed11a88d96e89b6835

Reply:
> And for this code line, I think I have found another good way to avoid calling :set(), I just call :get() with a new simple callback to query and update the stale/expired record. See https://github.com/Kong/kong/commit/a6029d085a833acf3dbf2b95645825e39b21f2c1

I think what you said about not calling `:update()` for this TTL-expired scenario is right. and here I cannot make sure that the mlcache will finnally call the L3 callback to query the dns record, it might get the L1/L2 cached result. So I have to query the dns server and `:set()` the result mannually.

Reply:
@chobits there is in our version `mlcache:renew`

Reply:
> @chobits there is in our version `mlcache:renew` (`mlcache:update` does almost nothing with new worker events - there is no poll): https://github.com/Kong/kong/blob/master/kong/resty/mlcache/init.lua#L786

Thanks for reminding me, I'll take a look at it.

Reply:
Hi after I change mlcache ipc shm to kong.worker_events, I have verified the `:set()` posted the lru invalidation event could be sent and received by other workers.